From b6a1e419033df1b8caff64b85b287c7f158a4762 Mon Sep 17 00:00:00 2001
From: Anson Huang <b20788@freescale.com>
Date: Mon, 2 Sep 2013 14:52:58 -0400
Subject: [PATCH 0189/1072] ENGR00277691 ARM: imx: Add speed grading check.

git://git.freescale.com/imx/linux-2.6-imx.git imx_3.10.17_1.0.0_beta
commit 87638b97be3cddad51acc818a4240412385571b7

For i.MX6Q/DL, different chips may have different
max speed of ARM defined in fuse map of speed_grading[1:0]:

2b'11: 1200000000Hz;
2b'10: 1000000000Hz;
2b'01: 850000000Hz; -- i.MX6Q Only, exclusive with 1GHz.
2b'00: 800000000Hz;

Need to read fuse data to set max speed of ARM.

Signed-off-by: Anson Huang <b20788@freescale.com>
Signed-off-by: Hongbo Zhong <hongbo.zhong@windriver.com>
---
 arch/arm/boot/dts/imx6q.dtsi   |    2 ++
 arch/arm/mach-imx/mach-imx6q.c |   26 +++++++++++++++++++++++---
 2 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/arch/arm/boot/dts/imx6q.dtsi b/arch/arm/boot/dts/imx6q.dtsi
index 93e75ff..b7c670b 100644
--- a/arch/arm/boot/dts/imx6q.dtsi
+++ b/arch/arm/boot/dts/imx6q.dtsi
@@ -29,6 +29,7 @@
 				/* kHz    uV */
 				1200000 1275000
 				996000  1250000
+				852000  1250000
 				792000  1150000
 				396000  975000
 			>;
@@ -36,6 +37,7 @@
 				/* ARM kHz  SOC-PU uV */
 				1200000       1275000
 				996000        1250000
+				852000        1250000
 				792000        1175000
 				396000        1175000
 			>;
diff --git a/arch/arm/mach-imx/mach-imx6q.c b/arch/arm/mach-imx/mach-imx6q.c
index 99ada11..fac290ae 100644
--- a/arch/arm/mach-imx/mach-imx6q.c
+++ b/arch/arm/mach-imx/mach-imx6q.c
@@ -221,8 +221,11 @@ static void __init imx6q_init_machine(void)
 #define OCOTP_CFG3			0x440
 #define OCOTP_CFG3_SPEED_SHIFT		16
 #define OCOTP_CFG3_SPEED_1P2GHZ		0x3
+#define OCOTP_CFG3_SPEED_1GHZ		0x2
+#define OCOTP_CFG3_SPEED_850MHZ		0x1
+#define OCOTP_CFG3_SPEED_800MHZ		0x0
 
-static void __init imx6q_opp_check_1p2ghz(struct device *cpu_dev)
+static void __init imx6q_opp_check_speed_grading(struct device *cpu_dev)
 {
 	struct device_node *np;
 	void __iomem *base;
@@ -240,11 +243,28 @@ static void __init imx6q_opp_check_1p2ghz(struct device *cpu_dev)
 		goto put_node;
 	}
 
+	/*
+	 * SPEED_GRADING[1:0] defines the max speed of ARM:
+	 * 2b'11: 1200000000Hz;
+	 * 2b'10: 1000000000Hz;
+	 * 2b'01: 850000000Hz; -- i.MX6Q Only, exclusive with 1GHz.
+	 * 2b'00: 800000000Hz;
+	 * We need to set the max speed of ARM according to fuse map.
+	 */
 	val = readl_relaxed(base + OCOTP_CFG3);
 	val >>= OCOTP_CFG3_SPEED_SHIFT;
-	if ((val & 0x3) != OCOTP_CFG3_SPEED_1P2GHZ)
+	if ((val & 0x3) < OCOTP_CFG3_SPEED_1P2GHZ)
 		if (opp_disable(cpu_dev, 1200000000))
 			pr_warn("failed to disable 1.2 GHz OPP\n");
+	if ((val & 0x3) < OCOTP_CFG3_SPEED_1GHZ)
+		if (opp_disable(cpu_dev, 996000000))
+			pr_warn("failed to disable 1 GHz OPP\n");
+	if (cpu_is_imx6q()) {
+		if ((val & 0x3) < OCOTP_CFG3_SPEED_850MHZ ||
+			(val & 0x3) == OCOTP_CFG3_SPEED_1GHZ)
+			if (opp_disable(cpu_dev, 852000000))
+				pr_warn("failed to disable 850 MHz OPP\n");
+	}
 
 put_node:
 	of_node_put(np);
@@ -266,7 +286,7 @@ static void __init imx6q_opp_init(struct device *cpu_dev)
 		goto put_node;
 	}
 
-	imx6q_opp_check_1p2ghz(cpu_dev);
+	imx6q_opp_check_speed_grading(cpu_dev);
 
 put_node:
 	of_node_put(np);
-- 
1.7.5.4

